extensions/amp-date-countdown/1.0/date-countdown.js

```diff
@@ -0,0 +1,245 @@
+};
+
+/**
+ * @param {!DateCountdownPropsDef} props
+ * @return {PreactDef.Renderable}
+ */
+export function DateCountdown({
+  endDate,
+  timeleftMs,
+  timestampMs,
+  timestampSeconds,
+  offsetSeconds = DEFAULT_OFFSET_SECONDS,
+  whenEnded = DEFAULT_WHEN_ENDED,
+  locale = DEFAULT_LOCALE,
+  biggestUnit = DEFAULT_BIGGEST_UNIT,
+  render,
+  children,
+}) {
+  useResourcesNotify();
+  const {playable} = useAmpContext();
+  const [epoch, setEpoch] = useState(
+    getEpoch(endDate, timeleftMs, timestampMs, timestampSeconds) +
+      offsetSeconds * DELAY
+  );
+  const [timeLeft, setTimeLeft] = useState(new Date(epoch) - new Date());
+
+  useEffect(() => {
+    const interval = setInterval(() => {
+      const newTimeLeft = new Date(epoch) - new Date();
+      setTimeLeft(() => newTimeLeft);
+      if (whenEnded === DEFAULT_WHEN_ENDED && newTimeLeft < 1000) {
+        clearInterval(interval);
+      }
+    }, DELAY);
+    return () => clearInterval(interval);
+  }, [playable]);
+
+  const data = Object.assign(
+    getYDHMSFromMs(timeLeft, biggestUnit),
+    getLocaleWord(locale)
+  );
+  return render(data, children);
+}
+
+/**
+ * Calculate the epoch time that this component should countdown to from
+ * one of multiple input options.
+ * @param {string|undefined} endDate
+ * @param {number|undefined} timeleftMs
+ * @param {number|undefined} timestampMs
+ * @param {number|undefined} timestampSeconds
+ * @return {number}
+ */
+function getEpoch(endDate, timeleftMs, timestampMs, timestampSeconds) {
+  let epoch;
+
+  if (endDate) {
+    epoch = Date.parse(endDate);
+  } else if (timeleftMs) {
+    epoch = Number(new Date()) + timeleftMs;
+  } else if (timestampMs) {
+    epoch = timestampMs;
+  } else if (timestampSeconds) {
+    epoch = timestampSeconds * 1000;
+  }
+
+  if (epoch === undefined) {
+    throw new Error(
+      `One of endDate, timeleftMs, timestampMs, timestampSeconds` +
+        `is required. ${NAME}`
+    );
+  }
+  return epoch;
+}
+
+/**
+ * Return an object with a label for 'years', 'months', etc. based on the
+ * user provided locale string.
+ * @param {string} locale
+ * @return {!Object}
+ */
+function getLocaleWord(locale) {
+  if (LOCALE_WORD[locale] === undefined) {
+    throwWarning(
+      `Invalid locale ${locale}, defaulting to ${DEFAULT_LOCALE}. ${NAME}`
+    );
+    locale = DEFAULT_LOCALE;
+  }
+  const localeWordList = LOCALE_WORD[locale];
+  return {
+    'years': localeWordList[0],
+    'months': localeWordList[1],
+    'days': localeWordList[2],
+    'hours': localeWordList[3],
+    'minutes': localeWordList[4],
+    'seconds': localeWordList[5],
+  };
+}
+
+/**
+ * Converts a time represented in milliseconds (ms) into a representation with
+ * days, hours, minutes, etc. and returns formatted strings in an object.
+ * @param {number} ms
+ * @param {string} biggestUnit
+ * @return {Object}
+ */
+function getYDHMSFromMs(ms, biggestUnit) {
+  /** @enum {number} */
```

Comment:
This is not a enum since this map is accessed by a variable key. This is `Object<string, number>`

Reply:
Thanks, updated in the hoisted const.